stdlib/public/SDK/Foundation/Foundation.swift.gyb

```diff
@@ -133,204 +133,133 @@ extension String : _ObjectiveCBridgeable {
-  }
-
-  @_semantics("convertToObjectiveC")
-  public func _bridgeToObjectiveC() -> NSNumber {
-    return _swift_Foundation_TypePreservingNSNumberWithInt(self)
-  }
 
-  public static func _forceBridgeFromObjectiveC(
-    _ x: NSNumber,
-    result: inout Int?
-  ) {
-    result = x.intValue
-  }
-
-  public static func _conditionallyBridgeFromObjectiveC(
-    _ x: NSNumber,
-    result: inout Int?
-  ) -> Bool {
-    self._forceBridgeFromObjectiveC(x, result: &result)
-    return true
-  }
+%{
+# The set of types we bridge to NSNumber using a Swift-type-preserving
+# subclass. Note that this doesn't include Bool or CGFloat, which require
+# special handling.
+bridgedNumberTypes = [
+  ('Int',     'int'),
+  ('Int64',   'int64'),
+  ('Int32',   'int32'),
+  ('Int16',   'int16'),
+  ('Int8',    'int8'),
+  ('UInt',    'uint'),
+  ('UInt64',  'uint64'),
+  ('UInt32',  'uint32'),
+  ('UInt16',  'uint16'),
+  ('UInt8',   'uint8'),
+  ('Float',   'float'),
+  ('Double',  'double'),
+]
+}%
+
+% for NumberType, accessorName in bridgedNumberTypes:
+
+@_silgen_name("_swift_Foundation_TypePreservingNSNumberWith${NumberType}")
+internal func _swift_Foundation_TypePreservingNSNumberWith${NumberType}(
+  _ value: ${NumberType}
+) -> NSNumber
 
-  public static func _unconditionallyBridgeFromObjectiveC(
-    _ source: NSNumber?
-  ) -> Int {
-    return source!.intValue
-  }
-}
+@_silgen_name("_swift_Foundation_TypePreservingNSNumberGetAs${NumberType}")
+internal func _swift_Foundation_TypePreservingNSNumberGetAs${NumberType}(
+  _ value: NSNumber
+) -> ${NumberType}
 
-extension UInt : _ObjectiveCBridgeable {
+extension ${NumberType} : _ObjectiveCBridgeable {
   public init(_ number: NSNumber) {
-    self = number.uintValue
+    self = number.${accessorName}Value
   }
 
   @_semantics("convertToObjectiveC")
   public func _bridgeToObjectiveC() -> NSNumber {
-    return _swift_Foundation_TypePreservingNSNumberWithUInt(self)
+    return _swift_Foundation_TypePreservingNSNumberWith${NumberType}(self)
   }
 
   public static func _forceBridgeFromObjectiveC(
     _ x: NSNumber,
-    result: inout UInt?
+    result: inout ${NumberType}?
   ) {
-    result = x.uintValue
-  }
-
-  public static func _conditionallyBridgeFromObjectiveC(
-    _ x: NSNumber,
-    result: inout UInt?
-  ) -> Bool {
-    self._forceBridgeFromObjectiveC(x, result: &result)
-    return true
-  }
-
-  public static func _unconditionallyBridgeFromObjectiveC(
-    _ source: NSNumber?
-  ) -> UInt {
-    return source!.uintValue
-  }
-}
-
-extension Float : _ObjectiveCBridgeable {
-  public init(_ number: NSNumber) {
-    self = number.floatValue
-  }
-
-  @_semantics("convertToObjectiveC")
-  public func _bridgeToObjectiveC() -> NSNumber {
-    return _swift_Foundation_TypePreservingNSNumberWithFloat(self)
-  }
+    // If the NSNumber instance preserved its Swift type, we only want to allow
+    // the cast if the type matches.
+    if let tag = _SwiftTypePreservingNSNumberTag(rawValue:
```

Comment:
I get the sneaky trick with the `NonSwift` value here, but it's really subtle. Could you document `_SwiftTypePreservingNSNumberTag` and it's C counterpart so nobody surprises themselves by trying to put `NonSwift` in `_SwiftTypePreservingNSNumberTag`?